engine/src/main/scala/cromwell/engine/workflow/lifecycle/finalization/WorkflowCallbackActor.scala

```diff
@@ -0,0 +1,226 @@
+  private lazy val maxRetriesWithDefault = maxRetries.getOrElse(defaultMaxRetries)
+
+  override def receive: Actor.Receive = LoggingReceive {
+    case PerformCallbackCommand(workflowId, requestedCallbackUri, terminalState, outputs, failures) =>
+      // If no uri was provided to us here, fall back to the one in config. If there isn't
+      // one there, do not perform a callback.
+      val callbackUri: Option[URI] = requestedCallbackUri.map(WorkflowCallbackConfig.createAndValidateUri).getOrElse(defaultCallbackUri)
+      callbackUri.map { uri =>
+        performCallback(workflowId, uri, terminalState, outputs, failures) onComplete {
+          case Success(_) =>
+            log.info(s"Successfully sent callback for workflow for workflow $workflowId in state $terminalState to $uri")
+            sendMetadata(workflowId, successful = true, uri)
+          case Failure(t) =>
+            log.warning(s"Permanently failed to send callback for workflow $workflowId in state $terminalState to $uri: ${t.getMessage}")
+            sendMetadata(workflowId, successful = false, uri)
+        }
+      }.getOrElse(())
+    case Broadcast(ShutdownCommand) | ShutdownCommand => context stop self
+    case other => log.warning(s"WorkflowCallbackActor received an unexpected message: $other")
+  }
+
+  private def makeHeaders: Future[List[HttpHeader]] = {
+    authMethod.toList.map(_.getAccessToken).map {
+      case Valid(header) => Future.successful(header)
+      case Invalid(err) => Future.failed(new RuntimeException(err.toString)) // TODO better error
+    }
+      .map(t => t.map(t => RawHeader("Authorization", s"Bearer $t")))
+      .traverse(identity)
+  }
+
+  private def performCallback(workflowId: WorkflowId, callbackUri: URI, terminalState: WorkflowState, outputs: CallOutputs, failures: List[String]): Future[Unit] = {
+    // Only send outputs if the workflow succeeded
+    val callbackPostBody = terminalState match {
+      case WorkflowSucceeded => CallbackMessage(workflowId.toString, terminalState.toString, Option(outputs.outputs.map(entry => (entry._1.name, entry._2))), None)
+      case WorkflowFailed => CallbackMessage(workflowId.toString, terminalState.toString, None, Option(failures))
+      case _ => CallbackMessage(workflowId.toString, terminalState.toString, None, None)
+    }
+    for {
+      entity <- Marshal(callbackPostBody).to[RequestEntity]
+      headers <- makeHeaders
+      request = HttpRequest(method = HttpMethods.POST, uri = callbackUri.toString, entity = entity).withHeaders(headers)
+      response <- withRetry(
+        () => sendRequestOrFail(request),
+        backoff = backoffWithDefault,
+        maxRetries = Option(maxRetriesWithDefault),
+        onRetry = err => log.warning(s"Will retry after failure to send workflow callback for workflow $workflowId in state $terminalState to $callbackUri : $err")
+      )
+      result <- {
+        response.entity.discardBytes()
```

Comment:
The documentation on this bizarre-sounding method is absolutely fascinating. Definitely seems correct here.